Contributing guidelines for new changelog generation (#1787)
matrix-org/matrix-js-sdk: CONTRIBUTING.md

it on github, and then create a pull request to ask us to pull your changes
into our repo (https://help.github.com/articles/using-pull-requests/)

We use GitHub's pull request workflow to review the contribution, and either
ask you to make any refinements needed or merge it and make them ourselves.

Things that should go into your PR description:
 * A changelog entry in the `Notes` section (see below)

>  * References to any bugs fixed by the change (in Github's `Fixes` notation)

Reviewer: ```suggestion
 * References to any bugs fixed by the change (in GitHub's `Fixes` notation)
```

 * Notes for the reviewer that might help them to understand why the change is
   necessary or how they might better review it.

Things that should *not* go into your PR description:
 * Any information on how the code works or why you chose to do it the way
   you did. If this isn't obvious from your code, you haven't written enough
   comments.

We rely on information in pull request to populate the information that goes

> into the changelogs our users see, both for the js-sdk itself and also for some

Reviewer: ```suggestion
into the changelogs our users see, both for the JS SDK itself and also for some
```

projects based on it. This is picked up from both labels on the pull request and

> the `Notes: ` annotation in the description. By default, the PR title will be

Reviewer 2: ```suggestion
the `Notes:` annotation in the description. By default, the PR title will be
```

Without this it gets rendered weirdly:
![Screenshot_20210714_210925](https://user-images.githubusercontent.com/25768714/125678976-bf961a30-a424-41d5-9d7f-fe6b7c547c75.png)

used for the changelog entry, but you can specify more options, as follows.

To add a longer, more detailed description of the change for the changelog:

*Fix llama herding bug*

```
Notes: Fix a bug (https://github.com/matrix-org/notaproject/issues/123) where the 'Herd' button would not herd more than 8 Llamas if the moon was in the waxing gibbous phase
```

For some PRs, it's not useful to have an entry in the user-facing changelog (this is

> the default for PRs labelled with `pr-internal`):

Reviewer: After the recent relabelling a few months back, we changes to using `Some-Words` casing in labels, so perhaps.

*Remove outdated comment from `Ungulates.ts`*
```
Notes: none
```

Sometimes, you're fixing a bug in a downstream project, in which case you want
an entry in that project's changelog. You can do that too:

*Fix another herding bug*
```
Notes: Fix a bug where the `herd()` function would only work on Tuesdays

> other-project notes: Fix a bug where the 'Herd' button only worked on Tuesdays

Reviewer 2: Do I understand correctly that the `other-project` should get replaced with a project name? If so, could it perhaps be made more clear (`<other-project>` or similar), it wasn't clear to me at the start

Author: Yeah - maybe a real-life example will be clearer: updated.

```

Projects that you can specify here are:
 * matrix-react-sdk
 * element-web
 * element-desktop

If your PR introduces a breaking change, use the `Notes` section in the same

> way, additionally adding the `pr-breaking` label (see below). There's no need

Reviewer: This might want a casing change as well.

There's no need to specify in the notes that it's a breaking change - this will
be added automatically based on the tag - but remember to tell the developer how
to migrate:

*Remove legacy class*

```
Notes: Remove legacy `Camelopard` class. `Giraffe` should be used instead.
```

> Other metadata can be added using tags.

Reviewer 3: Throughout: Are tags labels?

Author: Yes - maybe I've just never read the text that says 'labels' :/

>  * `pr-breaking`: A breaking change - adding this label will mean the change causes a *major* version bump.
>  * `pr-feature`: A new feature - adding this label will mean the change causes a *minor* version bump.
>  * `pr-bugfix`: A bugfix (in either code or docs).
>  * `pr-internal`: No user-facing changes, eg. code comments, CI fixes, refactors or tests. Won't have a changelog entry unless you specify one.

Reviewer: Along with casing change you're surely tired of by now... I wonder if there's some way to align (some of?) these with the [issue types](https://github.com/vector-im/element-web#triaging-issues)?

> * T-Defect: Bugs, crashes, hangs, vulnerabilities, or other reported problems
> * T-Enhancement: New features, changes in functionality, performance boosts, user-facing improvements
> * T-Task: Refactoring, enabling or disabling functionality, other engineering tasks
> * T-Other: Questions, user support, anything else

For example, a `T-Enhancement` issue could get a `T-Enhancement` PR that adds it, and then there's no need for a separate label at all.

```suggestion
 * `X-Breaking`: A breaking change - adding this label will mean the change causes a *major* version bump.
 * `T-Enhancement`: A new feature - adding this label will mean the change causes a *minor* version bump.
 * `T-Defect`: A bug fix (in either code or docs).
 * `T-Task`: No user-facing changes, eg. code comments, CI fixes, refactors or tests. Won't have a changelog entry unless you specify one.
```

Perhaps my OCD is going to far this time... 😅 Anyway, just a suggestion to perhaps fit this alongside the existing scheme (which mostly exists on the Element Web side only so far).

Author: Ah yep - this would also solve the problem of being able to add PR labels to issues (except X-Breaking I guess, although you could argue that there are some bug where fixing them would obviously necessitate a breaking change).
